### PR TITLE
Enable using mypy manually

### PR DESCRIPTION
**Pull request by MarcSkovMadsen**
In #7079, @gandhis1 contributes a typing improvement. In order to be able to review this PR I would like to run `mypy`.

This PR configures the project in such a way that you can

- run `pip install -e .[mypy]` to install the `mypy` dependencies
- run `mypy path/to/file.py` to check a specific file.

It removes the errors raised about missing stubs and dependencies that do not support typing.

### CODE REVIEW

**Comment by hoxbro** [comment]
I pushed some changes so we can see the result of type checking in the CI. I have set the CI to never fail, but this should be removed at some point.

File: pyproject.toml [hunk 1]
```diff
@@ -207,3 +217,52 @@ filterwarnings = [
     "ignore:unclosed file <_io.TextIOWrapper name='(/dev/null|nul)' mode='w':ResourceWarning", # OK
     "ignore:Deprecated in traitlets 4.1, use the instance .metadata dictionary directly", # OK (ipywidgets internal)
 ]
+
+[tool.mypy]
+python_version = "3.10"
```
Comment (hoxbro) [inline pyproject.toml:222]: I think this is automatically inferred by `requires-python`

File: pyproject.toml [hunk 2]
```diff
@@ -207,3 +217,52 @@ filterwarnings = [
...
+    "reacton.*",
+    "rpy2.*",
+    "scipy.*",
+    "setuptools_scm.*",
+    "snakeviz.*",
+    "streamz.*",
+    "textual.*",
+    "tranquilizer.*",
+    "vtk.*",
+
```
Comment (hoxbro) [inline pyproject.toml:266]:
```suggestion

```

File: .github/workflows/test.yaml [hunk 3]
```diff
@@ -257,3 +257,22 @@ jobs:
       - name: Test Unit
         run: |
           pixi run -e ${{ matrix.environment }} test-unit
+
+  type_test_suite:
+    name: lint:${{ matrix.environment }}:${{ matrix.os }}
```
Comment (hoxbro) [inline .github/workflows/test.yaml:262]:
```suggestion
    name: type:${{ matrix.environment }}:${{ matrix.os }}
```

File: .github/workflows/test.yaml [hunk 4]
```diff
@@ -257,3 +257,22 @@ jobs:
...
+        os: ["ubuntu-latest"]
+        environment: ["test-type"]
+    timeout-minutes: 120
+    steps:
+      - uses: holoviz-dev/holoviz_tasks/pixi_install@pixi
+        with:
+          environments: ${{ matrix.environment }}
+      - name: Test Type
+        run: |
+          pixi run -e ${{ matrix.environment }} type || echo "Failed"
```
Comment (hoxbro) [inline .github/workflows/test.yaml:278]:
```suggestion
          pixi run -e ${{ matrix.environment }} test-type || echo "Failed"
```

File: pixi.toml [hunk 5]
```diff
@@ -144,6 +145,22 @@ _install-ui = 'playwright install chromium'
...
+pandas-stubs = "*"
+types-bleach = "*"
+types-croniter = "*"
+types-Markdown = "*"
+types-psutil = "*"
+types-requests = "*"
+types-tqdm = "*"
+
+[feature.type.tasks]
+type = 'mypy panel'
```
Comment (hoxbro) [inline pixi.toml:162]:
```suggestion
test-type = 'mypy panel'
```

File: pixi.toml [hunk 6]
```diff
@@ -144,6 +145,22 @@ _install-ui = 'playwright install chromium'
 cmd = 'pytest panel/tests/ui --ui --browser chromium -n logical --dist loadgroup --reruns 3 --reruns-delay 10'
 depends_on = ["_install-ui"]
 
+# =============================================
+# ================== TYPES ====================
+# =============================================
+[feature.type.dependencies]
+mypy = "*"
+pandas-stubs = "*"
```
Comment (gandhis1) [inline pixi.toml:153]: `typing-extensions` usually useful to have for features that may have been backported from a newer version of Python

Comment (hoxbro) [reply pixi.toml:153]: Agreed, just copied the dependencies over from pyproject.

Comment (MarcSkovMadsen) [reply pixi.toml:153]: I've added typing-extensions.

File: pyproject.toml (hunk 1 above)
Comment (MarcSkovMadsen) [reply pyproject.toml]: I've removed the line.

File: pyproject.toml [hunk 7]
```diff
@@ -207,3 +218,50 @@ filterwarnings = [
     "ignore:unclosed file <_io.TextIOWrapper name='(/dev/null|nul)' mode='w':ResourceWarning", # OK
     "ignore:Deprecated in traitlets 4.1, use the instance .metadata dictionary directly", # OK (ipywidgets internal)
 ]
+
+[tool.mypy]
+namespace_packages = true
+explicit_package_bases = true
+mypy_path = ""
+exclude = []
```
Comment (gandhis1) [inline pyproject.toml:226]: I ran Mypy in the same way you were @MarcSkovMadsen and I got a bunch of additional errors on unrelated files. At least for now, maybe it's worth turning off the unrelated file errors?

```suggestion
exclude = []
follow_imports = "silent"
```

From https://mypy.readthedocs.io/en/stable/running_mypy.html#following-imports:

> If you are starting a new codebase and plan on using type hints from the start, we recommend you use either [--follow-imports=normal](https://mypy.readthedocs.io/en/stable/command_line.html#cmdoption-mypy-follow-imports) (the default) or [--follow-imports=error](https://mypy.readthedocs.io/en/stable/command_line.html#cmdoption-mypy-follow-imports). Either option will help make sure you are not skipping checking any part of your codebase by accident.

> If you are planning on adding type hints to a large, existing code base, we recommend you start by trying to make your entire codebase (including files that do not use type hints) pass under [--follow-imports=normal](https://mypy.readthedocs.io/en/stable/command_line.html#cmdoption-mypy-follow-imports). This is usually not too difficult to do: mypy is designed to report as few error messages as possible when it is looking at unannotated code.

> **Only if doing this is intractable, we recommend passing mypy just the files you want to type check and use [--follow-imports=silent](https://mypy.readthedocs.io/en/stable/command_line.html#cmdoption-mypy-follow-imports).** Even if mypy is unable to perfectly type check a file, it can still glean some useful information by parsing it (for example, understanding what methods a given object has). See [Using mypy with an existing codebase](https://mypy.readthedocs.io/en/stable/existing_code.html#existing-code) for more recommendations.

File: panel/io/cache.py [hunk 8]
```diff
@@ -427,7 +465,7 @@ async def wrapped_func(*args, **kwargs):
             return ret
     else:
         @functools.wraps(func)
-        def wrapped_func(*args, **kwargs):
+        def wrapped_func(*args: P.args, **kwargs: P.kwargs) -> R:
```
Comment (gandhis1) [inline panel/io/cache.py:468]: I think you may have inadvertently included this file?

File: pyproject.toml (hunk 7 above)
Comment (MarcSkovMadsen) [reply pyproject.toml:226]: I don't think we should change it with the argument that what we want to check in the pipeline is everything as that is the end goal for the Panel project to have no issues whether internal facing or external facing.

I believe for "the current effort" which is focused on outward facing api we should add the relevant settings via flags manually when running mypy manually. Makes sense?

File: panel/io/cache.py (hunk 8 above)
Comment (MarcSkovMadsen) [reply panel/io/cache.py]: Thx. I believe I've "undone" it by running `git checkout main -- panel/io/cache.py`.

File: pixi.toml [hunk 9]
```diff
@@ -144,6 +145,22 @@ _install-ui = 'playwright install chromium'
...
+types-bleach = "*"
+types-croniter = "*"
+types-Markdown = "*"
+types-psutil = "*"
+types-requests = "*"
+types-tqdm = "*"
```
Comment (hoxbro) [inline pixi.toml:159]:
```suggestion
types-tqdm = "*"
typing-extensions = "*"
```

**Review by gandhis1 — APPROVED** [review]
LGTM

**Review by hoxbro — APPROVED** [review]

**Review by philippjfr — APPROVED** [review]
